indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java

```diff
@@ -621,4 +653,96 @@ private static DimensionSchema createDimensionSchema(
       return usedSegments;
     }
   }
+
+  @VisibleForTesting
+  static class PartitionConfigurationManager
+  {
+    @Nullable
+    private final Long targetCompactionSizeBytes;
+    @Nullable
+    private final IndexTuningConfig tuningConfig;
+
+    PartitionConfigurationManager(@Nullable Long targetCompactionSizeBytes, @Nullable IndexTuningConfig tuningConfig)
+    {
+      this.targetCompactionSizeBytes = getValidTargetCompactionSizeBytes(targetCompactionSizeBytes, tuningConfig);
+      this.tuningConfig = tuningConfig;
+    }
+
+    @Nullable
+    IndexTuningConfig computeTuningConfig(List<Pair<QueryableIndex, DataSegment>> queryableIndexAndSegments)
+    {
+      if (!hasPartitionConfig(tuningConfig)) {
+        // Find IndexTuningConfig.targetPartitionSize which is the number of rows per segment.
+        // Assume that the segment size is proportional to the number of rows. We can improve this later.
+        final double avgRowsPerByte = queryableIndexAndSegments
+            .stream()
+            .mapToDouble(queryableIndexAndDataSegment -> {
+              final long numRows = queryableIndexAndDataSegment.lhs.getNumRows();
+              final long size = queryableIndexAndDataSegment.rhs.getSize();
+              return numRows / (double) size;
+            })
+            .average()
+            .orElseThrow(() -> new ISE("Cannot find targetPartitionSize because the input segment is empty"));
+        final int targetPartitionSize = (int) Math.round(avgRowsPerByte * targetCompactionSizeBytes);
+        Preconditions.checkState(targetPartitionSize > 0, "Negative targetPartitionSize[%s]", targetPartitionSize);
```

Comment:
Would it be better to use something like `Math.toIntExact()` to check for overflow here?

Reply:
Good point. Fixed.

Reply:
this will need to catch ArithmeticException instead of checking for negative